### PR TITLE
Merge `Hbc::Container` classes into `UnpackStrategy`.

### PR DESCRIPTION
**Pull request by reitermarkus**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

### CODE REVIEW

File: Library/Homebrew/system_command.rb [hunk 1]
```diff
@@ -151,10 +153,11 @@ def each_line_from(sources)
   end
 
   def result
-    Result.new(command,
-               processed_output[:stdout],
-               processed_output[:stderr],
-               processed_status.exitstatus)
+    output = @merged_output.each_with_object(stdout: "", stderr: "") do |(type, line), h|
+      h[type] << line
```
Comment (MikeMcQuaid) [inline Library/Homebrew/system_command.rb]: `hash` would be a preferable variable (assuming that's what it is)

File: Library/Homebrew/system_command.rb [hunk 2]
```diff
@@ -93,9 +92,9 @@ def sudo_prefix
   end
 
   def assert_success
-    return if processed_status&.success?
+    return if @status.success?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/system_command.rb]: Can `@status` ever be `nil`?

Comment (reitermarkus) [reply Library/Homebrew/system_command.rb]: I don't think so.

**Comment by ilovezfs** [comment]
This broke building boost from source.

```
iMac-TMP:Homebrew joe$ brew instal -dvs boost
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/boost.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/icu4c.rb
/usr/bin/sandbox-exec -f /private/tmp/homebrew20180725-62160-1q7fn1n.sb nice /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.7/bin/ruby -W0 -I /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.7/lib/ruby/gems/2.3.0/gems/did_you_mean-1.0.0/lib:/usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.7/lib/ruby/site_ruby/2.3.0:/usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.7/lib/ruby/site_ruby/2.3.0/x86_64-darwin9.0:/usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.7/lib/ruby/site_ruby/2.3.0/universal-darwin9.0:/usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.7/lib/ruby/site_ruby:/usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.7/lib/ruby/vendor_ruby/2.3.0:/usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.7/lib/ruby/vendor_ruby/2.3.0/x86_64-darwin9.0:/usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.7/lib/ruby/vendor_ruby/2.3.0/universal-darwin9.0:/usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.7/lib/ruby/vendor_ruby:/usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.7/lib/ruby/2.3.0:/usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.7/lib/ruby/2.3.0/x86_64-darwin9.0:/usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.7/lib/ruby/2.3.0/universal-darwin9.0:/usr/local/Homebrew/Library/Homebrew/cask/lib:/usr/local/Homebrew/Library/Homebrew -- /usr/local/Homebrew/Library/Homebrew/build.rb /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/boost.rb --verbose --debug
/usr/local/Homebrew/Library/Homebrew/build.rb (Formulary::FromPathLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/boost.rb
==> Downloading https://dl.bintray.com/boostorg/release/1.67.0/source/boost_1_67_0.tar.bz2
Already downloaded: /Users/joe/Library/Caches/Homebrew/boost-1.67.0.tar.bz2
==> Verifying boost-1.67.0.tar.bz2 checksum
==> /usr/bin/hdiutil imageinfo /Users/joe/Library/Caches/Homebrew/boost-1.67.0.tar.bz2
==> tar xf /Users/joe/Library/Caches/Homebrew/boost-1.67.0.tar.bz2 -C /private/tmp/d20180725-62162-aghybr
==> Downloading https://github.com/boostorg/lockfree/commit/12726cd.patch?full_index=1
Already downloaded: /Users/joe/Library/Caches/Homebrew/boost--patch-f165823d961a588b622b20520668b08819eb5fdc08be7894c06edce78026ce0a.patch
==> Verifying boost--patch-f165823d961a588b622b20520668b08819eb5fdc08be7894c06edce78026ce0a.patch checksum
==> Patching
/usr/local/Homebrew/Library/Homebrew/debrew.rb:11:in `raise'
MachO::MagicError: Unrecognized Mach-O magic: 0x46726f6d
1. raise
2. ignore
3. backtrace
4. irb
5. shell
Choose an action: /usr/local/Homebrew/Library/Homebrew/debrew.rb:56:in `exit'
/usr/local/Homebrew/Library/Homebrew/debrew.rb:56:in `choose'
/usr/local/Homebrew/Library/Homebrew/debrew.rb:112:in `block in debug'
/usr/local/Homebrew/Library/Homebrew/debrew.rb:111:in `loop'
/usr/local/Homebrew/Library/Homebrew/debrew.rb:111:in `debug'
/usr/local/Homebrew/Library/Homebrew/debrew.rb:14:in `rescue in raise'
/usr/local/Homebrew/Library/Homebrew/debrew.rb:10:in `raise'
/usr/local/Homebrew/Library/Homebrew/vendor/macho/macho.rb:35:in `open'
/usr/local/Homebrew/Library/Homebrew/unpack_strategy/executable.rb:11:in `can_extract?'
/usr/local/Homebrew/Library/Homebrew/unpack_strategy.rb:63:in `block in from_path'
/usr/local/Homebrew/Library/Homebrew/unpack_strategy.rb:62:in `each'
/usr/local/Homebrew/Library/Homebrew/unpack_strategy.rb:62:in `detect'
/usr/local/Homebrew/Library/Homebrew/unpack_strategy.rb:62:in `from_path'
/usr/local/Homebrew/Library/Homebrew/unpack_strategy.rb:80:in `detect'
/usr/local/Homebrew/Library/Homebrew/download_strategy.rb:50:in `stage'
/usr/local/Homebrew/Library/Homebrew/download_strategy.rb:173:in `stage'
/usr/local/Homebrew/Library/Homebrew/resource.rb:116:in `block in unpack'
/usr/local/Homebrew/Library/Homebrew/resource.rb:189:in `block in mktemp'
/usr/local/Homebrew/Library/Homebrew/mktemp.rb:55:in `block in run'
/usr/local/Homebrew/Library/Homebrew/mktemp.rb:55:in `chdir'
/usr/local/Homebrew/Library/Homebrew/mktemp.rb:55:in `run'
/usr/local/Homebrew/Library/Homebrew/resource.rb:188:in `mktemp'
/usr/local/Homebrew/Library/Homebrew/resource.rb:115:in `unpack'
/usr/local/Homebrew/Library/Homebrew/patch.rb:137:in `apply'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1104:in `each'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1104:in `patch'
/usr/local/Homebrew/Library/Homebrew/debrew.rb:26:in `block in patch'
/usr/local/Homebrew/Library/Homebrew/debrew.rb:92:in `debrew'
/usr/local/Homebrew/Library/Homebrew/debrew.rb:26:in `patch'
/usr/local/Homebrew/Library/Homebrew/build.rb:119:in `block (2 levels) in install'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1117:in `block in brew'
/usr/local/Homebrew/Library/Homebrew/formula.rb:2037:in `block (2 levels) in stage'
/usr/local/Homebrew/Library/Homebrew/utils.rb:564:in `with_env'
/usr/local/Homebrew/Library/Homebrew/formula.rb:2036:in `block in stage'
/usr/local/Homebrew/Library/Homebrew/resource.rb:120:in `block in unpack'
/usr/local/Homebrew/Library/Homebrew/resource.rb:189:in `block in mktemp'
/usr/local/Homebrew/Library/Homebrew/mktemp.rb:55:in `block in run'
/usr/local/Homebrew/Library/Homebrew/mktemp.rb:55:in `chdir'
/usr/local/Homebrew/Library/Homebrew/mktemp.rb:55:in `run'
/usr/local/Homebrew/Library/Homebrew/resource.rb:188:in `mktemp'
/usr/local/Homebrew/Library/Homebrew/resource.rb:115:in `unpack'
/usr/local/Homebrew/Library/Homebrew/resource.rb:93:in `stage'
/usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.7/lib/ruby/2.3.0/forwardable.rb:202:in `stage'
/usr/local/Homebrew/Library/Homebrew/formula.rb:2014:in `stage'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1112:in `brew'
/usr/local/Homebrew/Library/Homebrew/build.rb:113:in `block in install'
/usr/local/Homebrew/Library/Homebrew/utils.rb:564:in `with_env'
/usr/local/Homebrew/Library/Homebrew/build.rb:110:in `install'
/usr/local/Homebrew/Library/Homebrew/build.rb:191:in `<main>'
iMac-TMP:Homebrew joe$ git bisect bad
eccd20e2762955dc2de5e0b5af5fd982100bac63 is the first bad commit
commit eccd20e2762955dc2de5e0b5af5fd982100bac63
Author: Markus Reiter <me@reitermark.us>
Date:   Mon Jul 23 23:04:49 2018 +0200

    Merge `Hbc::Container` classes into `UnpackStrategy`.

:040000 040000 59326a6642808f980132df4e96cd6dbc56a91dfd 942ed8906461cd395e666ec18d1dfe41c560dea6 M	Library
```

**Comment by ilovezfs** [comment]
The line
```
==> /usr/bin/hdiutil imageinfo /Users/joe/Library/Caches/Homebrew/boost-1.67.0.tar.bz2
```
is also very distracting. That should never be relevant in formula build output.

**Comment by MikeMcQuaid** [comment]
I think it should be shown in verbose output but not otherwise.

**Comment by ilovezfs** [comment]
```
iMac-TMP:~ joe$ brew install -ds boost
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/boost.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/icu4c.rb
/usr/local/Homebrew/Library/Homebrew/build.rb (Formulary::FromPathLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/boost.rb
==> Downloading https://dl.bintray.com/boostorg/release/1.67.0/source/boost_1_67_0.tar.bz2
Already downloaded: /Users/joe/Library/Caches/Homebrew/boost-1.67.0.tar.bz2
==> hdiutil imageinfo /Users/joe/Library/Caches/Homebrew/boost-1.67.0.tar.bz2
==> tar xf /Users/joe/Library/Caches/Homebrew/boost-1.67.0.tar.bz2 -C /private/tmp/d20180725-80691-1osgrbc
^C
iMac-TMP:~ joe$ 
```
Just plain weird.

**Comment by ilovezfs** [comment]
How could it possibly be necessary to run `hdiutil info` on every source archive in formulae when exactly zero of them will be dmg, let alone display that useless execution?

**Comment by ilovezfs** [comment]
@reitermarkus e.g. can't the check for whether it's a tarball, etc. be moved up to occur much earlier so that it doesn't unnecessarily execute hdiutil info every time?

**Comment by reitermarkus** [comment]
`Tar` might be possible to be above `Dmg`.

**Comment by ilovezfs** [comment]
Thanks. Moving tar, jar, zip, etc. above dmg would be much appreciated as there's never a reason to be running `hdiutil imageinfo` on any source archive.

**Comment by ilovezfs** [comment]
Interestingly it even seems fond of running the command on git repositories.
```
git submodule update --init --recursive
==> hdiutil imageinfo /Users/joe/Library/Caches/Homebrew/mongrel2--git
```

**Comment by hn3000** [comment]
This breaks installation of nuget because the installed nuget.exe matches the magic number check for "MZ" (it's a .exe file), but the extraction fails with the error

    .../nuget-4.7.0.exe: no valid cabinets found

(I worked around it by changing the magic number check to not match "MZ".)

Can the check be improved? Or can a formula opt out of the automatic unpacking? (I did not see how they could ...)

**Comment by ilovezfs** [comment]
Also affects `paket`.

**Pull request merged**